쥬스메이커 [STEP2] Kiseok, 쥬봉이

jyubong opened the pull request:
@Tediousday93 
안녕하세요. step2 PR입니다. 잘부탁드립니다:)

### 코드를 작성할 때 고민되었던 점
- 화면 전환 방법
    - `modal` present 방식 : 주문화면과 재고수정화면은 엄연히 다른 역할이라고 생각해 네비게이션 처럼 연결성이 있다기 보다는 모달로 이목을 끌고 사용자가 빠르게 처리할 수 있도록 표현하기 위해서 사용했습니다. 
    - StockChangeViewController에서도 navigation bar사용을 위해 navigation controller와 연결합니다.

- 화면 사이 데이터 공유
    - `싱글톤 객체` 활용 : fruitStore를 싱글톤으로 구현하여 juiceMaker와 두 view controller가 접근, fruit의 재고를 변경시키고 데이터를 들고와 view를 최신화시킵니다.
    - 현 프로젝트에서는 동시에 싱글톤 객체에 접근하여 race condition을 발생시킬 일은 없을 것 같아 사용하였습니다.

### 해결이 되지 않은 점 & 조언을 얻고 싶은 부분
- 위에서 말씀드린 이유로 모달방식을 택하긴 하였지만, **아직도 어떤 경우에 어떤 방식을 선택해야할지 많이 헷갈립니다.**
- **싱글톤을 화면 데이터 공유에 사용하였는데, 이 부분이 적절한지 고민**이 됩니다. 현 프로젝트에서는 문제가 없지만 커지면 싱글톤으로 인해 문제가 생길 수도 있다고 생각합니다. 그래서 Notification이나 다른 화면 사이 데이터 공유 방식도 생각해 보았습니다.

carti1108 committed:
feat: 화면 전환 구현

네비게이션 방식으로 화면 전환 구현
재고수정 viewcontroller 추가

jyubong committed:
feat: view controller에 label과 button 연결

carti1108 committed:
feat: juiceOrderButton 통합, error enum 정의, error handling

jyubong committed:
feat: alert창 구현

주문시 성공과 실패시 alert 구현

carti1108 committed:
feat: 화면 전환 구현

재고 수정 버튼, alert창 예 버튼 클릭 시 화면 전환
수정완료 클릭 시 rootview로 돌아오기 구현
navigation 방식 구현

jyubong committed:
feat: 과일 재고 UI에 표시

fruitStore 싱글톤 객체의 재고 활용
private로 은닉화

carti1108 committed:
refactor: 화면 전환 방식 수정

네비게이션 방식에서 모달로 수정

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+
+import UIKit
+
+class JuiceMakerViewController: UIViewController {
```
뷰컨트롤러의 이름을 잘 바꿔주셨네요 ☺️
이 클래스는 상속할 가능성이 있을까요? 그렇지 않다면 붙여줄 수 있는 키워드가 있을 것 같아요 💭

jyubong replied:
상속 제한을 위해 class 앞에 `final` 키워드를 추가해주었습니다.
[애플 공식문서 prevent override](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/inheritance#Preventing-Overrides)

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+        pineappleStockLabel.text = String(fruitStore.fruits[.pineapple] ?? 0)
+        kiwiStockLabel.text = String(fruitStore.fruits[.kiwi] ?? 0)
+        mangoStockLabel.text = String(fruitStore.fruits[.mango] ?? 0)
+    }
```
뷰 생명주기 메서드 viewDidLoad 에서 label의 초기 text를 설정해주셨군요!
닐 병합 연산자도 잘 활용해주셨습니다. 👍👍

피드백에서 말씀드렸던 매직넘버 처리에 있어 0은 아래처럼 하드코딩을 피할 수 있습니다.
```swift
strawberryStockLabel.text = String(fruitStore.fruits[.strawberry] ?? .zero)
```

carti1108 replied:
0을 사용한 매직넘버 부분들은 .zero로 변경완료하였습니다!

jyubong replied:
피드백에서 얼럿을 만드는 메서드에서 활용할 수 있다고 하셔서 작성을 해보았는데... 이 부분을 알맞게 구현을 한것일까요?🥹

Tediousday93 replied:
정말 잘 바꿔주셨어요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+        pushToStockViewController()
+    }
+    
+    @IBAction func juiceOrderButtonTapped(_ sender: UIButton) {
+        var menu: JuiceMenu?
+        
+        switch sender.tag {
+        case 0:
+            menu = .strawberryBananaJuice
+        case 1:
+            menu = .strawberryJuice
+        case 2:
+            menu = .bananaJuice
+        case 3:
+            menu = .pineappleJuice
+        case 4:
+            menu = .mangoKiwiJuice
+        case 5:
+            menu = .kiwiJuice
+        case 6:
+            menu = .mangoJuice
+        default:
+            break
+        }
+        
+        makeJuice(of: menu)
+    }
```
tag를 활용해 버튼에 Action을 잘 연결해주셨네요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+            alertOutOfStock()
+        } catch {
+            print(error.localizedDescription)
+        }
```
오류처리와 얼럿을 잘 활용해주셨네요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+
+class JuiceMakerViewController: UIViewController {
+    
+    @IBOutlet weak var strawberryStockLabel: UILabel!
+    @IBOutlet weak var bananaStockLabel: UILabel!
+    @IBOutlet weak var pineappleStockLabel: UILabel!
+    @IBOutlet weak var kiwiStockLabel: UILabel!
+    @IBOutlet weak var mangoStockLabel: UILabel!
```
시간이 남으신다면 @IBOutlet 프로퍼티에서 볼 수 있는 특징에 대해 알아보셔도 좋겠네요!
1. `weak` 키워드를 사용하는 이유
2. 사용하지 말라고 했던 Implicitly Unwrapped Optional(암시적 추출 옵셔널)을 사용하는 이유

참고) @IBOutlet 프로퍼티에도 접근제어를 붙일 수 있습니다. 💭

jyubong replied:
#### @IBOutlet 프로퍼티
1. weak 키워드를 사용하는 이유
    - view와 view객체들 간에 강한 참조가 되어있다면 참조 카운트가 더 많이 늘어나고 순환 참조가 발생할 수 있음
2. Implicitly Unwrapped Optional(암시적 추출 옵셔널)을 사용하는 이유
    - 스토리보드를 통해 연결 해주었기 때문에 객체가 있다는 것은 확신하기 때문에 코드를 더 명확하고 편하게 작성하기 위함
    - 연결이 되지 않았다는 것을 강제적으로 알려주기 위함
    - **하지만 crash 발생을 방지하기 위해 안전하게 `optional로 변경`하라는 의견들도 많은데 변경하는 것이 좋을까요?**

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+    
+    private func alertOutOfStock() {
+        let alert = UIAlertController(title: nil, message: "재료가 모자라요. 재고를 수정할까요?", preferredStyle: .alert)
```
코드 한 줄이 많이 길어지거나 파라미터가 여러개인 경우에는 아래처럼 개행을 추가해주기도 합니다.
취향 차이이니 참고만 해주세요 😊
```suggestion
        let alert = UIAlertController(title: nil,
                                      message: "재료가 모자라요. 재고를 수정할까요?",
                                      preferredStyle: .alert)
```

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    
+    private func pushToStockViewController() {
+        guard let stockNavigationController = self.storyboard?.instantiateViewController(withIdentifier: "StockChangeNavigationController") as? UINavigationController else { return }
+        
```
개행 하나가 더 있는 것 같네요. 🤔

guard 문 같은 경우도 너무 길 경우에 개행을 추가하기도 합니다.
이 부분도 스타일이니 참고만 해주세요!
```suggestion
        guard let stockNavigationController = self.storyboard?.instantiateViewController(
            withIdentifier: "StockChangeNavigationController"
        ) as? UINavigationController else {
            return
        }
```

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+    
+    private func pushToStockViewController() {
+        guard let stockNavigationController = self.storyboard?.instantiateViewController(withIdentifier: "StockChangeNavigationController") as? UINavigationController else { return }
```
스토리보드에서 StockChangeViewController가 embed되어있던 navigation controller를 인스턴스화해주셨네요.
주어진 프로젝트 내용과 의미를 파악하는 것도 중요하죠! 너무 잘 해주셨습니다 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/StockChangeViewController.swift:
```diff
@@ -0,0 +1,37 @@
+
+import UIKit
+
+class StockChangeViewController: UIViewController {
```
여기도 마찬가지로 상속할 가능성이 없다면 적절한 키워드를 붙여주세요!

jyubong replied:
상속 제한을 위해 class 앞에 `final` 키워드를 추가해주었습니다.
[애플 공식문서 prevent override](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/inheritance#Preventing-Overrides)

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/Errors.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by Kiseok, 쥬봉이 on 2023/09/18.
+//
+
+enum Errors: Error {
```
어떤 Error인지 직관적으로 알 수 없을 것 같아요.
어느 타입 또는 메서드에서 발생하는 에러인지 알 수 있는 이름으로 수정해볼까요?

jyubong replied:
해당 에러는 현재 Stock에 대한 부분만 있어 Error 이름을 StockError로 변경하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,7 +8,7 @@
 class FruitStore {
     static let shared = FruitStore(initialStock: 10)
-    private var fruits: [Fruit: Int] = [:]
+    private(set) var fruits: [Fruit: Int] = [:]
```
setter만 private 접근수준으로 변경해주셨군요 👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,104 @@
+        guard let stockNavigationController = self.storyboard?.instantiateViewController(withIdentifier: "StockChangeNavigationController") as? UINavigationController else { return }
+        
+        
+        self.present(stockNavigationController, animated: true)
```
여기 present에는 self를 명시해주셨네요!
self를 명시하는 것도 기준을 잡아두고 일정하게 맞춰주시면 좋겠습니다 😊

carti1108 replied:
self 명시하는 부분 통일 완료하였습니다!

Tediousday93 reviewed (changes requested):
기석, 쥬봉이 안녕하세요 😊
step2 고생하셨습니다. 요구사항 정말 잘 지켜주셨네요!
전반적으로 타입 내부 코드들의 순서도 잘 맞춰주셔서 보기 수월했습니다.
접근제어도 잘 해주셨네요👏

질문이 생기시면 언제든 디스코드 DM 보내주세요 👍

### ✅ 고민했던 점
**화면 전환 방법**
* 화면의 역할과 맥락에 따른 화면 전환 방식에 대해 고민하셨군요!
* navigation controller도 잘 활용해주셨습니다.

**화면 사이 데이터 공유**
* 싱글톤 패턴을 사용하셨네요! 싱글톤 패턴의 장단점은 어떤 것이 있을까요? 싱글톤을 사용하지 않았을 때와 어떤 차이점이 있을까요?

</br>

### ✅ 해결이 되지 않은 점 / 조언을 얻고 싶은 부분
**모달 화면 전환 방식**
* 선택하신 이유가 합당하다고 생각합니다. 👍
* 작성해주신 PR 내용을 보면 화면 전환의 기준을 잘 이해하고 계신 것 같아요. ☺️

**싱글톤 패턴**
* 화면 간 데이터 공유 방식도 적절한 이유만 있다면 어느 것을 선택해도 문제는 없다고 생각합니다.
* 여러 가지 방법을 고민하시는 모습 좋습니다 👍
* Notification 외에도 또 다른 방법은 어떤 것들이 있는지 알아보시면 좋을 것 같아요!

</br>

### 🔸 그 외 피드백
뷰컨트롤러 내부에 매직리터럴/매직넘버가 있네요.
이를 처리하는 방법으로 **NameSpace**에 대해 알아보시면 좋겠습니다 😊

얼럿을 만드는 메서드에서 활용해볼 수 있겠네요 💪

jyubong committed:
refactor: 컨벤션 수정, final 키워드 추가, 메세지 namespace 추가

present -> self.present 통일
guard let 개행 변경
Int 0 -> .zero 수정
alert 메세지와 타이틀 매직리터럴 -> namespace 추가 후 변경

carti1108 commented:
주신 피드백에 대해서 아래와 같이 알아보았습니다!

#### singleton 사용시 장단점
- 장점
메모리 한 곳만 사용
객체 생성 시 시간 절약

- 단점
어디서든 접근이 가능 -> 어떤 객체와 연결되어 있는지 파악 어려움
인스턴스 하나만 참조하다보니 의존성이 강해짐
race condition 발생

싱글톤 사용하지 않는다면?
- 두 view controller가 같은 fruitStore로 접근하기 어려움
- fruit의 재고 확인을 위한 새로운 방법 생각 : Notification, KVO 등 다른 데이터 전달 방식 사

#### 데이터 전달 방법
- 다음 뷰컨트롤러의 인스턴스에 직접 대입 후 화면 전환
- KVO
- AppDelegate 객체 사용
    - UIApplication 객체로부터 생명 주기 관리를 위임받아 커스텀 코드를 처리하는 역할
    - 앱 전체를 통틀어 하나만 존재하여 여러 뷰 컨트롤러에서 모두 접근 가능, 앱 생명 주기와 함께함
    - 이 객체에 값을 저장하면 앱이 종료되지 않는 한 값을 계속 유지
- UserDefault 객체 사용
    - 이 객체를 사용하면 앱을 삭제하기 전까지는 저장된 값이 반영구적으로 유지됨

Tediousday93 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -17,18 +17,27 @@ class JuiceMakerViewController: UIViewController {
     private let juiceMaker: JuiceMaker = JuiceMaker()
     private let fruitStore: FruitStore = FruitStore.shared
     
+    enum Message {
+        static let yes: String = "예"
+        static let no: String = "아니오"
+        static let check: String = "확인"
+        static let none: String? = nil
+        static let success: String = " 나왔습니다! 맛있게 드세요!"
+        static let outOfStock: String = "재료가 모자라요. 재고를 수정할까요?"
+    }
```
Message enum을 중첩 타입으로 잘 만들어주셨습니다! 👏

다만, 몇 가지만 다시 고려해봅시다!
* 중첩 타입도 JuiceMakerViewController 외부에서 알 필요가 없다면 private 접근수준으로 설정하는 것이 알맞다고 생각되네요 💭
* `none` case 의 경우 nil을 할당해주셨는데, nil은 옵셔널 타입의 값이 없는 상태를 나타내는 키워드이므로 따로 네임스페이스에 빼두실 필요는 없을 것 같네요 😊

carti1108 replied:
private 키워드 추가 후 nil case는 제거 해주었습니다!

Tediousday93 reviewed (changes requested):
요구사항이 잘 반영되었네요 👏👏
고생하셨습니다!

여러가지 데이터 전달 방식은 앞으로 학습하게 될 기회가 있을거에요!
참조타입 인스턴스의 경우, 같은 인스턴스를 참조하기 위해 이니셜라이저를 통한 의존성 주입으로 해결하는 방법도 있답니다. 😊

메서드나 프로퍼티에 self를 붙여주는 부분의 기준은 잘 잡으셨을까요? 😊
저의 경우, 상속받은 메서드 등과 같이 타입 내부에 명시되어 있지 않은 메서드는 self를 붙여주는 편입니다.
항상 여러분들만의 이유와 기준을 정하실 수 있기를 바랍니다 :)

추가 코멘트만 수정해주시면 바로 머지하도록 하겠습니다 👍👍👍

carti1108 committed:
refactor: message 열거형 접근 제어 추가, nil 제거

Tediousday93 commented:
step2 고생하셨습니다. 👏👏👏
다음 스텝도 화이팅입니다! 💪
